typing/rec_check.ml

```diff
@@ -181,24 +197,25 @@ let classify_expression : Typedtree.expression -> sd =
     | Texp_apply _ ->
         Not_recursive
 
-    | Texp_for _
-    | Texp_constant _
     | Texp_new _
     | Texp_instvar _
     | Texp_tuple _
     | Texp_array _
-    | Texp_variant _
-    | Texp_setfield _
-    | Texp_while _
-    | Texp_setinstvar _
     | Texp_pack _
     | Texp_object _
     | Texp_function _
     | Texp_lazy _
-    | Texp_unreachable
     | Texp_extension_constructor _ ->
         Static
 
+    | Texp_for _
+    | Texp_constant _
+    | Texp_setfield _
+    | Texp_while _
+    | Texp_setinstvar _
+    | Texp_unreachable ->
+        Constant
```

Comment:
For readability I would suggest three groups, separated by a blank line: Texp_constant goes first (it is the canonical Constant construction), then all the `unit`-returning expressions, then `Texp_unreachable` which is weird. (Maybe there is a comment to add on the unreachable case?)